nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -0,0 +1,607 @@
+          nntrainer->valid_data->push_count, nntrainer->num_validation_samples);
+
+  ml_logd("<leaved>");
+  return 0;
+}
+
+int getSample(float **input, float **label, bool *last, void *user_data) {
+
+  auto data = reinterpret_cast<NNTrainer::InputTensorsInfo *>(user_data);
+
+  ml_logd("<called>");
+  ml_logd("(pop/push: %" PRId64 "/%" PRId64 ")", data->pop_count,
+          data->push_count);
+  pid_t pid = getpid();
+  pid_t tid = syscall(SYS_gettid);
+
+  ml_logd("<called>");
+  ml_logd("pid[%d], tid[%d]", pid, tid);
+  ml_logd("front:%d, rear:%d", data->queue_front, data->queue_rear);
+  ml_logd("num_inputs: %" PRId64 ", num_labels: %" PRId64 "", data->num_inputs,
+          data->num_labels);
+
+  unsigned int i = 0;
+  unsigned int idx = data->queue_front;
+  ml_logd("Delete, queue_front: %d", idx);
+
+  for (i = 0; i < data->num_inputs; i++) {
+    ml_logd("memcpy Addr %p, %p, size=%" PRId64 "\n", *(input + i),
+            data->tensor_data[idx].inputs[i], data->input_size[i]);
+    std::memcpy(*(input + i), data->tensor_data[idx].inputs[i],
+                data->input_size[i]);
+  }
+  for (i = 0; i < data->num_labels; i++) {
+    ml_logd("memcpy Addr %p, %p, size=%" PRId64 "", *(label + i),
+            data->tensor_data[idx].labels[i], data->label_size[i]);
+    std::memcpy(*(label + i), data->tensor_data[idx].labels[i],
+                data->label_size[i]);
+  }
+
+  data->pop_count++;
+  data->queue_count--;
+  data->queue_front = (data->queue_front + 1) % data->queue_size;
+
+  ml_logd("(pop/push: %" PRId64 "/%" PRId64 ")", data->pop_count,
+          data->push_count);
+
+  if (data->pop_count < data->total_num_samples) { // train or valid num samples
+    *last = false;
+  } else {
+    *last = true;
+    data->pop_count = 0;
+
+    std::random_device rd;
+    std::mt19937 g(rd());
+    std::shuffle(data->tensor_data.begin(), data->tensor_data.end(), g);
+  }
+
+  if (data->is_data_full_locked && data->queue_count > 0) {
+    data->data_full.notify_one();
+    ml_logd("send signal");
+  }
+  ml_logd("front:%d, rear:%d, filled:%d", data->queue_front, data->queue_rear,
+          data->queue_count);
+
+  /* epoch is complete */
+  if (data->pop_count == 0)
+    return 0;
+
+  /* to avoid dead lock, check is_data_full_locked */
+  if (!data->is_data_full_locked && data->queue_count == 0) {
+    ml_logd("locked, need to wait for more data");
+    std::unique_lock<std::mutex> lock(data->data_wait_lock);
+    data->is_data_wait_locked = TRUE;
+    data->data_wait.wait(lock);
+    ml_logd("unlocked, get data");
+  }
+
+  ml_logd("<leave>");
+
+  return 0;
+}
+
+void NNTrainer::NNTrainerTrain::createDataset() {
+
+  ml_logd("<called>");
+
+  train_data = std::make_unique<NNTrainer::InputTensorsInfo>(
+    num_training_samples, num_inputs, num_labels, tensors_inputsize);
+  valid_data = std::make_unique<NNTrainer::InputTensorsInfo>(
+    num_validation_samples, num_inputs, num_labels, tensors_inputsize);
+
```

Comment:
you might consider to add test_data set here.

Reply:
This part is not something I can decide.
@myungjoo said that evaluation is not done in the MLOps AI pipline.
Let's negotiate together.

Reply:
I'm not talking about the how we decide. NNTrainer provide the feature to test.. and it seems like it does not care about it. In my opinion, if you do not care about the test, then validation is also same here.